### PR TITLE
Fixed issues with generated OptionalExpand

### PR DESCRIPTION
**Pull request by pontusmelke**
Fixes include
- `CodeGenExpressions` never open tracers
- `HasLabel` do the right thing when node is `null`
- `AcceptVisitor` handles tracing for `Projection` and `ProduceResult`
- `OptionalExpand` updates boolean flag after passing predicates

### CODE REVIEW

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/codegen/ir/CheckingInstruction.scala [hunk 1]
```diff
@@ -20,23 +20,21 @@
 package org.neo4j.cypher.internal.compiler.v2_3.codegen.ir
 
 import org.neo4j.cypher.internal.compiler.v2_3.codegen.{CodeGenContext, MethodStructure}
-import org.neo4j.cypher.internal.compiler.v2_3.codegen.ir.expressions.CodeGenExpression
 
-case class Project(opName: String, projections: Seq[CodeGenExpression], parent: Instruction) extends Instruction {
+/**
+ * Generates instruction for for updating a provided flag before creating generating the inner instruction
```
Comment (lutovich) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/codegen/ir/CheckingInstruction.scala]: Either 'creating' or 'generating' should be removed

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/codegen/LogicalPlanConverter.scala [hunk 2]
```diff
@@ -19,324 +19,355 @@
...
+        val fromNodeVar = context.getVariable(optionalExpand.from.name)
+        val typeVar2TypeName = optionalExpand.types.map(t => context.namer.newVarName() -> t.name).toMap
+        val opName = context.registerOperator(optionalExpand)
+
+        //name of flag to check if results were yielded
+        val yieldFlag = context.namer.newVarName()
+
+        //wrap inner instructions with predicates
+        val instructionWithPredicates = optionalExpand.predicates
+          .reverseMap(ExpressionConverter.createPredicate(_)(context)).foldLeft[Instruction](CheckingInstruction(action, yieldFlag)) {
```
Comment (lutovich) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/codegen/LogicalPlanConverter.scala]: This is a very complicated line, can we please split it

File: community/cypher/cypher-compiler-2.3/src/main/java/org/neo4j/cypher/internal/compiler/v2_3/codegen/CompiledConversionUtils.java [hunk 3]
```diff
@@ -28,7 +28,7 @@
 // Class with static methods used by compiled execution plans
 public abstract class CompiledConversionUtils
 {
-    public static boolean isPropertyValueTrue( Object value )
+    public static boolean coerceToPredicate( Object value )
```
Comment (lutovich) [inline community/cypher/cypher-compiler-2.3/src/main/java/org/neo4j/cypher/internal/compiler/v2_3/codegen/CompiledConversionUtils.java:31]: Maybe `coerceToBoolean`? `coerceToPredicate` sounds like something that should return a function

**Comment by lutovich** [comment]
Made couple minor comments. Looks good otherwise

**Comment by systay** [comment]
Pushed manually

**Pull request closed without merging**